web/packages/teleport/src/services/queryHelpers.ts

```diff
@@ -0,0 +1,138 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as reactQuery from '@tanstack/react-query';
+import {
+  type DataTag,
+  type DefaultError,
+  type UseMutationOptions,
+  type UseMutationResult,
+  type UseQueryOptions,
+  type UseQueryResult,
+} from '@tanstack/react-query';
+
+export type MutationHook<
+  TVariables = void,
+  TData = unknown,
+  TError = DefaultError,
+> = (
+  options?: Omit<UseMutationOptions<TData, TError, TVariables>, 'mutationFn'>
+) => UseMutationResult<TData, TError, TVariables>;
+
+export function wrapMutation<
+  TVariables = void,
+  TData = unknown,
+  TError = DefaultError,
+>(
+  mutationFn: (variables: TVariables) => Promise<TData>
+): MutationHook<TVariables, TData, TError> {
+  return function wrappedMutation(
+    options?: Omit<UseMutationOptions<TData, TError, TVariables>, 'mutationFn'>
+  ) {
+    return reactQuery.useMutation<TData, TError, TVariables>({
+      mutationFn,
+      ...options,
+    });
+  };
+}
+
+export type QueryHook<
+  TData = unknown,
+  TVariables = void,
+  TError = DefaultError,
+> = (
+  variables?: TVariables,
+  options?: Omit<UseQueryOptions<TData, TError>, 'queryKey' | 'queryFn'>
+) => UseQueryResult<TData, TError>;
+
+export interface WrappedQuery<
+  TData = unknown,
+  TVariables = void,
+  TError = DefaultError,
+> {
+  createQueryKey: (variables?: TVariables) => DataTag<string[], TData, TError>;
+  queryKey: DataTag<string[], TData, TError>;
+  useQuery: QueryHook<TData, TVariables, TError>;
+}
+
+type SignalOnlyQueryFn<TData> = (signal: AbortSignal) => Promise<TData>;
+type VariablesQueryFn<TData, TVariables> = (
+  variables: TVariables,
+  signal: AbortSignal
+) => Promise<TData>;
+
+type QueryFn<TData, TVariables> = TVariables extends void
+  ? SignalOnlyQueryFn<TData>
+  : VariablesQueryFn<TData, TVariables>;
+
+export function wrapQuery<
+  TData = unknown,
+  TVariables = void,
+  TError = DefaultError,
+>(
+  queryKey: string[],
+  queryFn: QueryFn<TData, TVariables>
+): WrappedQuery<TData, TVariables, TError> {
+  return {
+    queryKey: queryKey as DataTag<string[], TData, TError>,
+    createQueryKey(variables?: TVariables) {
+      const key = [...queryKey];
+
+      if (variables) {
+        key.push(JSON.stringify(variables));
+      }
+
+      return key as DataTag<string[], TData, TError>;
+    },
+    useQuery: function wrappedQuery(
+      variables?: TVariables,
+      options?: Omit<UseQueryOptions<TData, TError>, 'queryKey' | 'queryFn'>
+    ) {
+      const key = [...queryKey];
+
+      if (variables) {
+        key.push(JSON.stringify(variables));
+      }
+
+      return reactQuery.useQuery({
+        queryKey: key,
+        queryFn: ({ signal }) => callQueryFn(queryFn, variables, signal),
+        ...options,
+      });
+    },
+  };
+}
+
+function isSignalOnlyQueryFn<TData = unknown, TVariables = void>(
```

Comment:
I'm open to seeing how this will work out. In the past I was trying to solve a similar problem in `useAsync`. Ultimately I gave up because I didn't want to prescribe how `signal` is supposed to be passed to the underlying function.

Expecting service methods to be in the shape of `(args, signal?) => { … }` or `(signal?) => { … }` seems reasonable though. I think it might not work for the gRPC client in Connect (see the snippet below), but we can have another wrapper specifically for it. It could also handle cloneable signals.

https://github.com/gravitational/teleport/blob/63e269a7569e38befb8965fde8212587c52d3144/web/packages/teleterm/src/ui/Vnet/vnetContext.tsx#L159-L167